GCEAgent/accounts.go

```diff
@@ -0,0 +1,237 @@
+//  Copyright 2017 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"os/user"
+	"reflect"
+	"strings"
+	"time"
+
+	"../logger"
+
+	"github.com/dchest/uniuri"
+)
+
+var (
+	regName         = "PublicKeys"
+	pwChars         = []byte(`abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789()~!@#$%^&*-+=|\{}[]:;<>,.?/`)
+	pwLgth          = 15
+	accountDisabled = false
+)
+
+type windowsKeyJSON struct {
+	Email    string
+	ExpireOn string
+	Exponent string
+	Modulus  string
+	UserName string
+}
+
+var badExpire []string
+
+func (k windowsKeyJSON) expired() bool {
+	t, err := time.Parse(time.RFC3339, k.ExpireOn)
+	if err != nil {
+		if !containsString(k.ExpireOn, badExpire) {
+			logger.Errorln("Error parsing time:", err)
+			badExpire = append(badExpire, k.ExpireOn)
+		}
+		return true
+	}
+	return t.Before(time.Now())
+}
+
+func (k windowsKeyJSON) createOrResetPwd() (*credsJSON, error) {
+	pwd := uniuri.NewLenChars(pwLgth, pwChars)
```

Comment:
I'm actually not sure why I settled on this library, it's not exactly unique in its generation